src/storage/v2/replication/replication_client.hpp

```diff
@@ -61,85 +66,80 @@ class ReplicaStream {
+    // that this and other transaction replication functions can only be
+    // called from a one thread stands)
+    if (State() != replication::ReplicaState::REPLICATING) {
+      return;
+    }
+    if (replica_stream_->IsDefunct()) return;
+    try {
+      callback(*replica_stream_);  // failure state what if not streaming (std::nullopt)
+    } catch (const rpc::RpcFailedException &) {
+      return replica_state_.WithLock([](auto &state) { state = replication::ReplicaState::MAYBE_BEHIND; });
+      LogRpcFailure();
+    }
+  }
+
   // Return whether the transaction could be finalized on the replication client or not.
-  [[nodiscard]] bool FinalizeTransactionReplication();
-
- protected:
-  virtual void RecoverReplica(uint64_t replica_commit) = 0;
-
-  auto GetStorage() -> Storage * { return storage_; }
-  auto LastCommitTimestamp() const -> uint64_t;
-  void InitializeClient();
-  void HandleRpcFailure();
-  void TryInitializeClientAsync();
-  void TryInitializeClientSync();
-  void FrequentCheck();
-
-  std::string name_;
-  communication::ClientContext rpc_context_;
-  rpc::Client rpc_client_;
-  std::chrono::seconds replica_check_frequency_;
-
-  std::optional<ReplicaStream> replica_stream_;
-  memgraph::replication::ReplicationMode mode_{memgraph::replication::ReplicationMode::SYNC};
-
-  utils::SpinLock client_lock_;
-  // This thread pool is used for background tasks so we don't
-  // block the main storage thread
-  // We use only 1 thread for 2 reasons:
-  //  - background tasks ALWAYS contain some kind of RPC communication.
-  //    We can't have multiple RPC communication from a same client
-  //    because that's not logically valid (e.g. you cannot send a snapshot
-  //    and WAL at a same time because WAL will arrive earlier and be applied
-  //    before the snapshot which is not correct)
-  //  - the implementation is simplified as we have a total control of what
-  //    this pool is executing. Also, we can simply queue multiple tasks
-  //    and be sure of the execution order.
-  //    Not having mulitple possible threads in the same client allows us
-  //    to ignore concurrency problems inside the client.
-  utils::ThreadPool thread_pool_{1};
-  std::atomic<replication::ReplicaState> replica_state_{replication::ReplicaState::INVALID};
-
-  utils::Scheduler replica_checker_;
-  Storage *storage_;
+  [[nodiscard]] bool FinalizeTransactionReplication(Storage *storage);
 
-  memgraph::replication::ReplicationEpoch const *repl_epoch_;
+  void TryCheckReplicaStateAsync(Storage *storage);  // TODO Move back to private
+ private:
+  void RecoverReplica(uint64_t replica_commit, memgraph::storage::Storage *storage);
+
+  void CheckReplicaState(Storage *storage);
+  void LogRpcFailure();
+  void TryCheckReplicaStateSync(Storage *storage);
+  void FrequentCheck(Storage *storage);
+
+  ::memgraph::replication::ReplicationClient &client_;
+  std::optional<ReplicaStream>
```

Comment:
TODO: make granted streams local variables, rather than stored here

Reply:
Not sure this is actually possible/advisable. The stream is a crucial part of the state machine.

Reply:
Can be done, will be done later on.